feature: wifi segment

mikesigs opened the pull request:
### Prerequisites

- [x] I have read and understood the `CONTRIBUTING` guide
- [x] The commit message follows the [conventional commits][cc] guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes/features)

### Description

This is a WiFi segment to display details of the current WiFi connection. I've only tested it so far on Windows, but the `netsh` command that makes it work there, also works in WSL. So I am pretty sure it will work in WSL too, but obviously need to test.

Any suggestions for how to do that? I was thinking I'd just re-clone the repo in WSL and install another Go environment there, build it, and test.

Also, for the Linux functionality, I tried to use the Dev Container in this project, but after much Googling it turns out virtual environments like that don't actually have a WiFi network interface. The data should be in `/proc/net/wireless`, but as I don't have a physical Linux machine to test with, I cannot develop this functionality. So I'm hoping it's cool to submit this with just Windows and WSL for now. And perhaps a kind contributor could help with the Linux and Darwin code!

Please let me know what you think of the PR. I am very happy to make any changes you request.

Here's a sample of what it looks like with the default template (yes that's my actual SSID):

![image](https://user-images.githubusercontent.com/811177/142752960-9b45254f-1b76-4f2b-afc8-b409d3f75c8d.png)


[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

JanDeDobbeleer commented on .gitignore:
```diff
@@ -158,3 +158,5 @@ Output/
 # images
 *.png
+
+local/
```
I don't think this is related somewhere, right?

mikesigs replied:
I put a local/ folder in the root to contain a theme for testing, then I also outputed the binary there. I can remove it. But how do you handle testing a new segment? Do you just use your jandedobbeleer theme all the time?

JanDeDobbeleer replied:
I have my own config in the home folder. I adjust that one while debugging (and leave the tasks file assumed unchanged at all times).

JanDeDobbeleer commented on src/segment_wifi.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+const (
+	ConnectedIcon    Property = "connected_icon"
```
These need to disappear in favor of a template that can handle swapping the icon with a simple if statement. We're trying to get rid of as muych icons as we can bny leveraging templates (you actually already have the logic in `defaultTemplate`)

mikesigs replied:
Do you mean, it'd be up to the user to use the template to prefix with their desired icon? And I should just bake this into the default template like `{{if .Connectected}}\uFAA8{{else}}\uFAA9{{end}}`?

JanDeDobbeleer replied:
Exactly!

mikesigs replied:
Cool. This is much cleaner. Got rid of an oddity in the tests I didn't like too.

JanDeDobbeleer commented on src/segment_wifi.go:
```diff
@@ -0,0 +1,118 @@
+			return false
+		}
+
+		regex := regexp.MustCompile(`(.+) : (.+)`)
```
please use the helper functions found in `regex.go`
- [getCompiledRegex](https://github.com/JanDeDobbeleer/oh-my-posh/blob/accf46420ff2924e863f5a76179abc92b2212535/src/regex.go#L15)
- [findNamedRegexMatch](https://github.com/JanDeDobbeleer/oh-my-posh/blob/accf46420ff2924e863f5a76179abc92b2212535/src/regex.go#L35)

mikesigs replied:
I wound up just removing the regex altogether and using `strings.Split(" : ")` instead

JanDeDobbeleer commented on src/segment_wifi.go:
```diff
@@ -0,0 +1,118 @@
+	// If in Linux, who wnows. Gonna need a physical linux machine to develop this since the wlan interface isn't available in wsl
+	// But possible solution is to cat /proc/net/wireless, and if that file doesn't exist then don't show the segment
+	// In Windows and in wsl you can use the command `netsh.exe wlan show interfaces`. Note the .exe is important for this to worw in wsl.
+	if w.env.getPlatform() == windowsPlatform || w.env.isWsl() {
```
Invert the if statement to have an early return rather than a full block indented.

JanDeDobbeleer commented on src/segment_wifi.go:
```diff
@@ -0,0 +1,118 @@
+		}
+	}
+
+	return true
```
shouldn't this be `false`?

mikesigs replied:
If you get past all the other gates in the method (which return false) then at this point it should be true. I refactored this method with your early exit suggestion, as well as pulled out the actual parsing logic into a new method. Makes it much more readable.

JanDeDobbeleer commented on src/segment_wifi_test.go:
```diff
@@ -0,0 +1,194 @@
+	return fmt.Sprintf(netshString, args.state, args.ssid, args.radioType, args.authentication, args.channel, args.receiveRate, args.transmitRate, args.signal)
+}
+
+func bootStrapEnvironment(args *wifiArgs) *wifi {
```
We've stepped away from this in favor of using test cases, see [here](https://github.com/JanDeDobbeleer/oh-my-posh/blob/accf46420ff2924e863f5a76179abc92b2212535/src/segment_aws_test.go#L9) for example.

mikesigs replied:
I tried to see how to write these tests using test cases. But I really didn't like the structure of it. Two things I couldn't really grasp:

1. How do you handle mocking dependencies? 
2. How do you handle test cases that don't assert just the simple output of the function?

Too much logic and indirection in testing code makes tests much harder to maintain imo. This is your repository, so I'll check out the link you provided and do my best to restructure.

JanDeDobbeleer replied:
I can always help out. Haven't found a downside as it only moves the bootstrap inside the for loop rather than have x functions. It also helps extending use-cases a lot faster.

mikesigs replied:
I will take a kick at it. I've pushed up the other changes you requested and suggested. This is the last piece.

mikesigs replied:
Done. And I see the light :) It's way cleaner.

JanDeDobbeleer reviewed (changes requested):
Am I correct in thinking this now only works in Windows? If so, please also add that context to the docs, we can always extend Unix (I could at least). Adding a few architectural requests, the test might be the most impactful changes, everything else should be feasible.

JanDeDobbeleer commented on .vscode/settings.json:
```diff
@@ -15,5 +15,27 @@
     "-s"
-  ]
+  ],
+  "workbench.colorCustomizations": {
```
this needs to be removed, it's a bit too much :-)

mikesigs replied:
Sorry, this is from the Peacock extension. I will remove it.

JanDeDobbeleer replied:
It was pretty colorful, I admit 😄

mikesigs commented:
So general question: How should I go about testing this in WSL? So far I've only run the code in Windows, but I have a high level of confidence it'll work in WSL as it's just going to use the `netsh.exe` form of the command, rather than trying to do it the Linux way (wsl itself doesn't have a wlan device). Can I just call `/path/to/my/compiled/oh-my-posh.exe` from wsl, or should I recompile the binary in wsl itself and use that?

mikesigs commented:
Also, this branch is now out of date with main. Do you prefer I rebase onto main? Or update with a merge commit from main?

mikesigs committed:
feat(wifi): added wifi segment for windows and wsl

mikesigs commented:
This is my first time using Go. It was a fun experience. I think things are pretty much ready now, other than proper WSL testing and the lack of *nix support.

JanDeDobbeleer commented:
> Do you prefer I rebase onto main

@mikesigs if you can that would be awesome, otherwise I just squash this.

JanDeDobbeleer commented on docs/docs/segment-wifi.md:
```diff
@@ -0,0 +1,50 @@
+
+Show details about the currently connected WiFi network.
+
+_Note: Currently only supports Windows and WSL. Pull requests for Darwin and Linux support are welcome :)_
```
You can turn this into an [admonition](https://docusaurus.io/docs/markdown-features/admonitions):

```markdown
:::info
Note...
:::
```

mikesigs replied:
Done

JanDeDobbeleer commented on themes/schema.json:
```diff
@@ -1685,6 +1686,38 @@
+            "properties": {
+              "properties": {
+                "properties": {
+                  "connected_icon": {
```
we still need to remove the icons, right?

mikesigs replied:
and done

JanDeDobbeleer commented:
2 comments, really great work. For your first attempt at go, I'm super proud of what you did here.

mikesigs committed:
docs: added wifi docs and schema

mikesigs commented:
Alrighty. I rebased and squashed most of the commits. Thanks for taking all of this time to review the PR each day. Your quick feedback has been really appreciated! I'm still curious how I should go about testing this in WSL, and if you have any recommendations. Also, I should say thanks for creating this tool. I used oh-my-posh v2 for far far too long. I had a bunch of local customizations I had made to the PowerShell module and was afraid to upgrade. Fortunately, you had basically implemented everything I had done in v2, and the v3 upgrade has been pretty seamless. 

The only thing lacking from my previous customizations was displaying the git branch name in the window/tab title. Perhaps I'll take a crack at that next, if it's something you think would be useful.

JanDeDobbeleer commented:
> displaying the git branch name in the window/tab title

@mikesigs that's a tricky one. Although we could find a way I guess. We'd need to cache the context.

mikesigs committed:
chore: fixing linting errors in build
